marlowe-contracts/test/Spec/Marlowe/Analysis.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE NumericUnderscores  #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module Spec.Marlowe.Analysis
+where
+
+import Data.Either (isRight)
+import Data.Maybe (isJust, isNothing)
+import Language.Marlowe
+import Language.Marlowe.Analysis.FSSemantics (warningsTrace)
+import Marlowe.Contracts.Options
+import Marlowe.Contracts.Swap
+import Test.Tasty
+import Test.Tasty.HUnit
+
+tests :: TestTree
+tests = testGroup "Marlowe Contract"
+    [ testCase "Swap test" swapTest
+    , testCase "Sequentially Swap twice test" sequentiallySwapTwiceTest
+    , testCase "Concurrently Swap twice test" concurrentlySwapTwiceTest
+    , testCase "Covered Call test" coveredCallTest
+    , testCase "Option test" optionTest
+    ]
+
+testWarnings :: Bool -> Contract -> IO ()
+testWarnings b contract = do
+  result <- warningsTrace contract
+  assertBool "Analysis ok" $ isRight result && either (const False) f result
+  where
+    f | b = isJust
+      | otherwise = isNothing
+
+testNoWarnings :: Contract -> IO ()
+testNoWarnings = testWarnings False
+
+testExpectedWarnings :: Contract -> IO ()
+testExpectedWarnings = testWarnings True
```

Comment:
This one was trickier than I thought. I think it is easier to understand like this:
```suggestion
hasWarnings :: Contract -> IO Bool
hasWarnings contract = do
  result <- warningsTrace contract
  case result of
    Left errDesc -> assertFailure ("Analysis failed: " ++ show errDesc)
    Right counterExample -> return $ isJust counterExample

testNoWarnings :: Contract -> Assertion
testNoWarnings contract =
  assertBool "Has no warnings" =<< not <$> hasWarnings contract

testExpectedWarnings :: Contract -> Assertion
testExpectedWarnings contract =
  assertBool "Has warnings" =<< hasWarnings contract
```
And also output in case of error should be clearer. Hopefully I didn't make any mistake